source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -147,23 +226,124 @@ void CacheFilter::onBody(Buffer::InstancePtr&& body) {
+  ASSERT(request_headers_, "injectValidationHeaders precondition unsatisfied: request_headers_ "
+                           "does not point to the RequestHeadersMap of the current request");
+  const Http::HeaderEntry* etag_header =
+      lookup_result_->headers_->get(Http::CustomHeaders::get().Etag);
+  const Http::HeaderEntry* last_modified_header =
+      lookup_result_->headers_->get(Http::CustomHeaders::get().LastModified);
+
+  if (etag_header) {
+    absl::string_view etag = etag_header->value().getStringView();
+    request_headers_->setReferenceKey(Http::CustomHeaders::get().IfNonMatch, etag);
+  }
+  if (last_modified_header) {
+    absl::string_view last_modified = last_modified_header->value().getStringView();
+    request_headers_->setCopy(Http::CustomHeaders::get().IfUnmodifiedSince, last_modified);
+  }
 }
+
+void CacheFilter::encodeCachedResponse() {
+  ASSERT(lookup_result_, "encodeCachedResponse precondition unsatisfied: lookup_result_ "
+                         "does not point to a cache lookup result");
+
+  response_has_trailers_ = lookup_result_->has_trailers_;
+  const bool end_stream = (lookup_result_->content_length_ == 0 && !response_has_trailers_);
+  // TODO(toddmgreer): Calculate age per https://httpwg.org/specs/rfc7234.html#age.calculations
+  lookup_result_->headers_->addReferenceKey(Http::Headers::get().Age, 0);
+
+  // Set appropriate response flags and codes
+  Http::StreamFilterCallbacks* callbacks_ =
+      filter_state_ == FilterState::Encoding
+          ? static_cast<Http::StreamFilterCallbacks*>(encoder_callbacks_)
+          : static_cast<Http::StreamFilterCallbacks*>(decoder_callbacks_);
+
+  callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::ResponseFromCacheFilter);
+  callbacks_->streamInfo().setResponseCodeDetails(
+      CacheResponseCodeDetails::get().ResponseFromCacheFilter);
+
+  // If the filter is encoding, 304 response headers and cached headers are merged in encodeHeaders
+  // If the filter is decoding, we need to serve response headers from cache directly
+  if (filter_state_ == FilterState::Decoding) {
+    // decoder_callbacks_->encodeHeaders() will invoke CacheFilter::encodeHeaders, skip it
+    skip_encoding_ = true;
+    decoder_callbacks_->encodeHeaders(std::move(lookup_result_->headers_), end_stream);
+  }
+
+  if (lookup_result_->content_length_ > 0) {
+    remaining_body_.emplace_back(0, lookup_result_->content_length_);
+    getBody();
+  } else if (response_has_trailers_) {
+    getTrailers();
+  }
+}
+
+void CacheFilter::finishedEncodingCachedResponse() {
+  if (filter_state_ == FilterState::Decoding) {
+    // Nothing needs to be done if a cache hit was found while decoding a request
+    return;
+  }
+  if (encode_cached_response_state_ == EncodeCachedResponseState::IterationStopped) {
+    // Filter iteration was stopped waiting for encoding cached response to finish -- continue
+    encoder_callbacks_->continueEncoding();
```

Comment:
Any chance `continueEncoding()` can return before finishing in some kind of paused state? (I have no idea, just checking)

Reply:
My (very limited) understanding is that `continueEncoding` only sets some state, so that the event loop on this worker thread picks up where it left in encoding in one of the next iterations. I may be (very) wrong though.
@htuch can you help us out here?

Reply:
When you continue encoding, the encoding proceeds to later filters. They are able to stop iteration (just like the cache filter). I'm not sure if that fits the definition of "paused" here.

Reply:
What I was concerned about was if `continueEncoding` can less than fully do the encoding that is being called finished at the end of the function, mainly because `continueEncoding` doesn't sound very final.

Reply:
I think there is some confusion here. 
When the cache filter is validating a response and receives a 304 response (not modified) it needs to add the cached response body to the encoding stream. If the cache is asynchronous, the filter pauses the encoding stream by returning `StopIteration` from `encodeHeaders` until all the data is added to the stream.
After the body is fetched from cache and is added to the encoding stream, the filter calls `continueEncoding` to tell the connection manager to continue the encoding stream (undo the `StopIteration` returned by `encodeHeaders`).

Reply:
@fredlas does this answer your question?

Reply:
Yes, thank you. I think my confusion demonstrates a need for a comment here.